lib/graphql/stitching/shaper.rb

```diff
@@ -0,0 +1,84 @@
+# typed: false
+# frozen_string_literal: true
+
+module GraphQL
+  module Stitching
+    class Shaper
+      def initialize(supergraph:, document:)
+        @supergraph = supergraph
+        @document = document
+      end
+
+      def perform(raw_result)
+        if ! raw_result.key?("errors") || raw_result["errors"].empty?  # TODO should do null cleanup for errors too
+          munge_entry(raw_result["data"], @document.operation.selections, @supergraph.schema.query)
+        end
+
+        # hate doing a second pass, but cannot remove _STITCH_ fields until the fragements are processed
+        clean_entry(raw_result["data"])
+
+        raw_result
+      end
+
+      private
+
+      def munge_entry(entry, selections, parent_type)
+        selections.each do |node|
+          if node.respond_to?(:name) && node&.name == "__typename"
```

Comment:
This check should move down into the `GraphQL::Language::Nodes::Field` as that is the only type to which the `__typename` corner-case applies. I assume you need to special-case this still because the GraphQL Schema doesn't list it as a possible field?